waspc/waspls/src/Wasp/LSP/TypeInference.hs

```diff
@@ -0,0 +1,132 @@
+module Wasp.LSP.TypeInference
+  ( -- * Inferred types for CST locations
+    inferTypeAtLocation,
+
+    -- * Lower level pieces
+    ExprPath,
+    ExprKey (..),
+    findExprPathAtLocation,
+    findTypeForPath,
+  )
+where
+
+import Control.Monad (guard)
+import Data.Foldable (find)
+import qualified Data.HashMap.Strict as M
+import qualified Wasp.Analyzer.Parser.CST as S
+import Wasp.Analyzer.Parser.CST.Traverse (Traversal)
+import qualified Wasp.Analyzer.Parser.CST.Traverse as T
+import Wasp.Analyzer.StdTypeDefinitions (stdTypes)
+import Wasp.Analyzer.Type (Type)
+import qualified Wasp.Analyzer.Type as Type
+import Wasp.Analyzer.TypeDefinitions (DeclType (dtBodyType), getDeclType)
+import Wasp.LSP.Syntax (lexemeAt)
+
+-- | Infer the type at a location in the CST, or 'Nothing' if type could not be
+-- determined.
+inferTypeAtLocation :: String -> Traversal -> Maybe Type
+inferTypeAtLocation src location = case findExprPathAtLocation src location of
+  Just path -> findTypeForPath path
+  _ -> Nothing
+
+-- | A "path" through wasp expressions to a certain location.
+--
+-- === __Example__
+-- For the code
+--
+-- @
+-- app todoApp {
+--   auth: {
+--     usernameAndPassword: |
+--                          ^
+--   }
+-- }
+-- @
+--
+-- The path to the cursor would be @[Decl "app", Key "auth", Key "usernameAndPassword"]@.
+type ExprPath = [ExprKey]
+
+-- | A single step of an 'ExprPath'.
+data ExprKey
```

Comment:
Why `ExprKey` name? Some ideas for a nicer name: `ExprPathStep`, `ExprPathLevel`.

Reply:
It's used in places as more than just part of an ExprPath -> but you say in comments it is `A single step of an 'ExprPath'`. If it is used as something more, maybe we should adjust comments.

Also, the need to say `A single step of an ExprPath` in comments is what kind of tells me it should probably be named that way, `ExprPathStep`.

If you are sure "key" is the right term, ok, but the only way I have seen it in our code so far was as dict key, and its usage in ExprPath / Signature is new and unknown to me, therefore devoid of meaning.